src/analysis/seminorm.lean

```diff
@@ -609,6 +609,24 @@ begin
   exact (map_add_le_add p _ _).trans (add_le_add hy₁ hy₂)
 end
 
+lemma sub_mem_ball (p : seminorm 𝕜 E) (x₁ x₂ y : E) (r : ℝ) :
+  x₁ - x₂ ∈ p.ball y r ↔ x₁ ∈ p.ball (x₂ + y) r :=
+by simp_rw [mem_ball, sub_sub]
+
+/- Can this be done more neatly with `seminorm.ball_comp`?
+Alternatively, can this be done using/imitating the ball-addition-lemmas
+in `normed_space/pointwise`? -/
+/-- The image of a ball under addition with a singleton is another ball. -/
+lemma singleton_add_ball (p : seminorm 𝕜 E) :
+  (λ (z : E), x + z) '' p.ball y r = p.ball (x + y) r :=
```

Comment:
As you noticed, this already exists for norms, so you should match the statements and names: `vadd_ball` (or `ball_add_singleton` if you don't want to deal with `add_opposite`).

Reply:
Thanks for the remark! I was thinking about this for a bit, and in principle, this would be a straightforward consequence of the ```vadd_ball``` lemma from ```normed/group/add_torsor``` if one just makes use that the seminorm induces a pseudometric structure on E and a torsor structure of E on itself. Would that be a desirable direction to go into? To me, that sounds like enough work to be its own pull request, in any case!

Reply:
I think that is a reasonable thing to do and it shouldn't be hard. I think you only have to use [add_group_seminorm.to_seminormed_add_group](https://leanprover-community.github.io/mathlib_docs/analysis/normed/group/basic.html#add_group_seminorm.to_seminormed_add_group), and then the typeclass system knows that a `seminormed_add_comm_group` is a `normed_add_torsor` on itself. But that can definitely wait for another PR if you prefer.

Reply:
I've been trying around to make your suggestion work and it looks really promising, but I am severely struggling in trying to make the identity ```‖x - y‖ = ⇑p (x - y)``` work out. All the simp_rws and unfolds in the world don't seem to get me anywhere. 

``` 
lemma vadd_ball (p : seminorm 𝕜 E) :
  x +ᵥ p.ball y r = p.ball (x +ᵥ y) r :=
begin
  haveI := add_group_seminorm.to_seminormed_add_comm_group p.to_add_group_seminorm,
  have : p.ball x r = metric.ball x r,
  { sorry }
end
```

And even when I force the two to be equal, with ```have : _inst.norm = p := sorry,```, it still doesn't work: The following ```refl``` does not solve it.

```
lemma vadd_ball (p : seminorm 𝕜 E) :
  x +ᵥ p.ball y r = p.ball (x +ᵥ y) r :=
begin
  haveI := add_group_seminorm.to_seminormed_add_comm_group p.to_add_group_seminorm,
  have : p.ball x r = metric.ball x r,
  have : _inst.norm = p := sorry,
  {
    unfold ball at *,
    unfold metric.ball at *,
    simp_rw dist_eq_norm at *,
    rw this, 
    -- refl,
  }
end
```

What's going on here? You'll probably see it immediately, otherwise I'll ask on Zulip. Just building the MWE sounds like work, too :)

Reply:
Ah yes I see the problem! You should use `letI` instead of `haveI`. The reason is that there is data in a `seminormed_add_comm_group` instance (namely, the norm), and `haveI` (like `have`) forgets the definition. Essentially, you're saying "equip `E` with *some* semnormed structure (which exists because [...])" instead of "equip `E` with *the* seminormed structure coming from [...]".

Reply:
Aaah, thank you so much! With that piece of knowledge, it was indeed very easy :)